source/server/listener_manager_impl.cc

```diff
@@ -561,23 +562,58 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
   }
 }
 
+void ListenerManagerImpl::stopListener(Network::ListenerConfig& listener,
+                                       std::function<void()> callback) {
+  const uint64_t listener_tag = listener.listenerTag();
+  const auto workers_pending_stop = std::make_shared<std::atomic<uint32_t>>(workers_.size());
+  for (const auto& worker : workers_) {
+    worker->stopListener(listener, [this, callback, workers_pending_stop, listener_tag]() {
+      if (--(*workers_pending_stop) == 0) {
+        server_.dispatcher().post([this, callback, listener_tag]() {
+          for (auto& listener : active_listeners_) {
```

Comment:
What is the case where we are stopping active listeners without draining? During shutdown I guess? (More comments please)

Reply:
I've factored non-sharable logic to callsites, should be easier to understand what's going on now